File: src/memory.zig
const std = @import("std");
const os = @import("root");

const c = os.c;
const mmio = os.mmio;
const scheduler = os.scheduler;
const framebuffer = os.framebuffer;

const assert = std.debug.assert;
const BitSet = os.datastruct.BitSet;

extern var _start: u8;
extern var __rodata_start: u8;
extern var __data_start: u8;
extern var __bss_end: u8;

// Creates a series of set bits. Useful when defining bit masks to use in
// communication with hardware.
fn ones(comptime count: u16) std.math.IntFittingRange(0, (1 << count) - 1) {
    comptime {
        return (1 << count) - 1;
    }
}

const mmu_bits = struct {
    // See D5-2726
    const block: u64 = 0b1 << 1;
    const pte: u64 = 0b1 << 1;
    const valid: u64 = 0b1 << 0;

    // This flag is managed by software and from what I understand it
    // handles page faults, similar to the "present" bit in x64
    const access: u64 = 0x1 << 10;

    // Executable permission bits
    const nx: u64 = 0x1 << 54; // disable execution for non-priveleged
    const px: u64 = 0x1 << 53; // disable execution for priveleged

    // D4-2735
    const rw_el1: u64 = 0b00 << 6;
    const r_el1: u64 = 0b10 << 6;
};

export var kernel_memory_map_pmd: [4096]u8 align(4096) = @bitCast([4096]u8, pmd_initial);
const pmd_initial = map: {
    @setEvalBranchQuota(2000);

    var pmd = [1]u64{0} ** 512;

    const mmio_base = mmio.MMIO_BASE >> 21;

    // Entry 0 of this map gets overwritten at boot-time, but that is
    // not important to the logic here.
    for (pmd) |*slot, i| {
        var descriptor: u64 = i << 21;

        descriptor |= mmu_bits.valid;
        descriptor |= mmu_bits.access;
        descriptor |= mmu_bits.nx;
        descriptor |= mmu_bits.px;

        const mair_bits = if (i < mmio_base) c.MT_NORMAL_NC_FLAGS else c.MT_DEVICE_nGnRnE;
        descriptor |= mair_bits << 2;

        slot.* = descriptor;
    }

    break :map pmd;
};

export var kernel_memory_map_pte: [4096]u8 align(4096) = map: {
    @setEvalBranchQuota(2000);

    var pte = [1]u64{0} ** 512;

    const mair_bits = c.MT_NORMAL_NC_FLAGS << 2;

    for (pte) |*slot, i| {
        var descriptor: u64 = i << 12;

        descriptor |= mmu_bits.valid;
        descriptor |= mmu_bits.pte;
        descriptor |= mmu_bits.access;
        descriptor |= mair_bits;
        descriptor |= mmu_bits.nx;

        // The stack starts at 0x80000 and should not be executable
        if (i < 128) {
            descriptor |= mmu_bits.px;
        }

        slot.* = descriptor;
    }

    break :map @bitCast([4096]u8, pte);
};

// These will go into bss and get initialized at runtime to all zeros;
// Page D5-2717 in the reference manual says that when bit[0] is set
// to 0, the entire descriptor is invalid, so that means that most of
// the initialization work is done for us when the kernel BSS is initialized
//
// They need to be initialized at runtime because we can't know the final address
// of these objects in the binary until we get to the linker, where it's already
// too late.
export var kernel_memory_map_pgd: [4096]u8 align(4096) = [1]u8{0} ** 4096;
export var kernel_memory_map_pud: [4096]u8 align(4096) = [1]u8{0} ** 4096;

fn addressPteBits(ptr: anytype) usize {
    return (@ptrToInt(ptr) >> 12) & 511;
}

pub fn initProtections() void {
    const exe_begin = addressPteBits(&_start);
    const exe_end = addressPteBits(&__rodata_start);
    const ro_end = addressPteBits(&__data_start);

    const pte = @ptrCast(*volatile [512]u64, &kernel_memory_map_pte);

    for (pte[exe_begin..exe_end]) |*slot| {
        slot.* |= mmu_bits.r_el1;
    }

    for (pte[exe_end..ro_end]) |*slot| {
        slot.* |= mmu_bits.r_el1 | mmu_bits.px;
    }

    for (pte[ro_end..]) |*slot| {
        slot.* |= mmu_bits.px;
    }

    asm volatile (
        \\tlbi vmalle1is
        \\DSB ISH
        \\isb
    );
}

// ----------------------------------------------------------------------------
//
//                          Global Memory Allocator
//
// ----------------------------------------------------------------------------
//
// This section implements a global memory allocator using the Buddy Allocator
// algorithm.

const virtual_base: u64 = 0xffff000000000000;
pub inline fn physicalAddress(ptr: anytype) u64 {
    const a = @ptrToInt(ptr);
    return a - virtual_base;
}

pub inline fn kernelPtr(comptime T: type, address: u64) T {
    return @intToPtr(T, address + virtual_base);
}

const class_count = 12;
const FreeBlock = extern struct {
    class: u64 align(4096),
    next: ?*@This(),
    prev: ?*@This(),
};

const ClassInfo = struct {
    freelist: ?*align(4096) FreeBlock,
    buddies: BitSet,
};

const BuddyInfo = struct {
    buddy: u64,
    bitset_index: u64,
};

var free_memory: u64 = 0;

// Note: these are only ever used for safety
var usable_pages: BitSet = undefined;
var free_pages: BitSet = undefined;

// NOTE: The smallest size class is 4kb.
//
// Also: right now, the buddy bitset has no meaning for the largest size class,
// since it has no buddy; we may want to change that later
const buddy_max = std.mem.alignForward(os.memory_size, 4096 << class_count);
const buddy_end_page = buddy_max / 4096;

var class0 = [1]u64{0} ** (buddyInfo(buddy_end_page, 0).bitset_index / 64);
var class1 = [1]u64{0} ** (buddyInfo(buddy_end_page, 1).bitset_index / 64);
var class2 = [1]u64{0} ** (buddyInfo(buddy_end_page, 2).bitset_index / 64);
var class3 = [1]u64{0} ** (buddyInfo(buddy_end_page, 3).bitset_index / 64);
var class4 = [1]u64{0} ** (buddyInfo(buddy_end_page, 4).bitset_index / 64);
var class5 = [1]u64{0} ** (buddyInfo(buddy_end_page, 5).bitset_index / 64);
var class6 = [1]u64{0} ** (buddyInfo(buddy_end_page, 6).bitset_index / 64);
var class7 = [1]u64{0} ** (buddyInfo(buddy_end_page, 7).bitset_index / 64);
var class8 = [1]u64{0} ** (buddyInfo(buddy_end_page, 8).bitset_index / 64);
var class9 = [1]u64{0} ** (buddyInfo(buddy_end_page, 9).bitset_index / 64);
var class10 = [1]u64{0} ** (buddyInfo(buddy_end_page, 10).bitset_index / 64);

// This initial value is valid for an initial set-up where all memory is used.
var classes = classes: {
    var classes_value: [class_count]ClassInfo = undefined;

    const class_arrays = [_][]u64{
        &class0, &class1, &class2,  &class3,
        &class4, &class5, &class6,  &class7,
        &class8, &class9, &class10,
    };

    for (class_arrays) |array, i| {
        classes_value[i] = .{
            .buddies = BitSet.initRaw(array, buddyInfo(buddy_end_page, i).bitset_index),
            .freelist = null,
        };
    }

    classes_value[11] = .{
        .buddies = BitSet.initRaw(&.{}, 0),
        .freelist = null,
    };

    break :classes classes_value;
};

// TODO: add a bigger-arena freelist, for allocations that go beyond the max
// class size?

pub fn initAllocator() void {
    // The `classes` object starts in a valid state, whose meaning is "all data
    // is allocated." This means we can safely just release the pages that
    // are usable.
    const begin = @ptrCast([*]align(4096) u8, @alignCast(4096, &__bss_end));
    const end = mmio.MMIO_BASE;

    const page_count = (end - @ptrToInt(begin)) / 4096;
    releasePagesImpl(begin, @intCast(u32, page_count));
}

fn buddyInfo(page: u64, class: u6) BuddyInfo {
    // assert(is_aligned(page, 1 << class));

    return BuddyInfo{
        .buddy = page ^ (@as(u64, 1) << class),
        .bitset_index = page >> (class + 1),
    };
}

fn addToFreelist(page: u64, class: u64) void {
    // assert(is_aligned(page, 1 << class));

    const block = kernelPtr(*FreeBlock, page * 4096);
    const info = &classes[class];

    block.class = class;
    block.prev = null;
    block.next = info.freelist;

    if (info.freelist) |head| {
        head.prev = block;
    }

    info.freelist = block;
}

fn removeFromFreelist(page: u64, class: u64) void {
    // assert(class < CLASS_COUNT);
    // assert(is_aligned(page, 1 << class));

    const block = kernelPtr(*FreeBlock, page * 4096);
    const info = &classes[class];

    if (block.next) |next| next.prev = block.prev;
    if (block.prev) |prev| prev.next = block.next else {
        assert(info.freelist == block);

        info.freelist = block.next;
    }
}

pub fn allocPages(requested_count: u32, best_effort: bool) error{OutOfMemory}![]align(4096) u8 {
    if (requested_count == 0) return &[0]u8{};

    scheduler.preemptDisable();
    defer scheduler.preemptEnable();

    const result = found_class: {
        const Result = struct { class: u6, freelist: *FreeBlock, count: u64 };

        const min_class = std.math.log2_int_ceil(u32, requested_count);
        for (classes[min_class..]) |class, i| {
            const free = class.freelist orelse continue;

            break :found_class Result{
                .count = requested_count,
                .freelist = free,
                .class = @intCast(u6, i + min_class),
            };
        }

        if (!best_effort) return error.OutOfMemory;

        var i = min_class;
        while (i > 0) {
            i -= 1;

            const free = classes[i].freelist orelse continue;

            break :found_class Result{
                .count = @as(u64, 1) << @truncate(u6, i),
                .freelist = free,
                .class = i,
            };
        }

        return error.OutOfMemory;
    };

    const count = result.count;
    const class = result.class;
    const freelist = result.freelist;

    // Pop from the freelist
    assert(freelist.class == class);
    classes[class].freelist = freelist.next;
    if (classes[class].freelist) |head| {
        head.prev = null;
    }

    const size = count * 4096;
    const buf = @ptrCast([*]align(4096) u8, freelist)[0..size];
    free_memory -= size;

    const addr = physicalAddress(buf.ptr);
    const begin = addr / 4096;
    const end = begin + count;
    _ = end;

    // TODO: safety checks
    // assert(all are usable);
    // assert(all are free);
    // set free pages to not free

    if (class != class_count - 1) {
        const index = buddyInfo(begin, class).bitset_index;
        assert(classes[class].buddies.isSet(index));
        classes[class].buddies.unset(index);
    }

    var i = class;
    var page = begin;
    var remaining = count;

    while (remaining > 0 and i > 0) : (i -= 1) {
        const child_class = i - 1;
        const info = &classes[child_class];
        const child_size = @as(u64, 1) << child_class;
        const index = buddyInfo(page, child_class).bitset_index;

        assert(info.buddies.isSet(index) == false);

        if (remaining > child_size) {
            remaining -= child_size;
            page += child_size;
            continue;
        }

        addToFreelist(page + child_size, child_class);
        info.buddies.set(index);

        if (remaining == child_size) break;
    }

    std.log.info("allocated {*}..{*}", .{ buf.ptr, buf.ptr + buf.len });
    return buf;
}

fn releasePagesImpl(data: [*]align(4096) u8, count: u32) void {
    scheduler.preemptDisable();
    defer scheduler.preemptEnable();

    const addr = physicalAddress(data);
    // assert(addr == align_down(addr, _4KB));

    const begin = addr / 4096;
    const end = begin + count;

    // assert(BitSet__get_all(MemGlobals.usable_pages, begin, end));
    // assert(!BitSet__get_any(MemGlobals.free_pages, begin, end));

    var free_page = begin;
    page: while (free_page < end) : (free_page += 1) {
        // TODO should probably do some math here to not have to iterate over every
        // page in data

        var page = free_page;
        for (classes[0..(class_count - 1)]) |*info, class| {
            // assert(is_aligned(page, 1 << class));
            const buds = buddyInfo(page, @truncate(u6, class));

            const buddy_is_free = info.buddies.isSet(buds.bitset_index);
            info.buddies.setValue(buds.bitset_index, !buddy_is_free);

            if (!buddy_is_free) {
                addToFreelist(page, class);
                continue :page;
            }

            removeFromFreelist(buds.buddy, class);
            page = std.math.min(page, buds.buddy);
        }

        // assert(is_aligned(page, 1 << (CLASS_COUNT - 1)));
        addToFreelist(page, class_count - 1);
    }

    free_memory += count * 4096;
    // BitSet__set_range(MemGlobals.free_pages, begin, end, true);
}

pub fn releasePages(data: [*]align(4096) u8, count: u32) void {
    releasePagesImpl(data, count);
    std.log.info("freed {*}..{*}", .{ data, data + count * 4096 });
}

File: src/memory.c
#include "memory.h"
#include "asm.h"
#include "bootboot.h"
#include "init.h"
#include "page_tables.h"
#include <basics.h>
#include <bitset.h>
#include <macros.h>
#include <types.h>

#define CLASS_COUNT 12
extern const u8 code_begin;
extern const u8 code_end;
extern const u8 bss_end;

typedef struct {
  MMapEnt *data;
  s64 count;
} MMap;

typedef struct memory__FreeBlock {
  struct memory__FreeBlock *next;
  struct memory__FreeBlock *prev;
  s64 class;
} FreeBlock;

typedef struct {
  s64 buddy;
  s64 bitset_index;
} BuddyInfo;

typedef struct {
  FreeBlock *freelist;
  BitSet buddies;
} ClassInfo;

typedef struct {
  const s64 size;
  const char *const suffix;
} MemSizeFormat;

static struct {
  s64 free_memory;

  // NOTE: This is used for checking correctness. Maybe its not necessary?
  BitSet usable_pages;
  BitSet free_pages;

  // NOTE: The smallest size class is 4kb.
  ClassInfo classes[CLASS_COUNT];
} MemGlobals;

Bump InitAlloc;

static void *alloc_from_entries(MMap mmap, s64 size, s64 align);

// Allocate `count` physically contiguous pages.
// Then:
//    if (exact == true), return NULL buffer.
//    if (exact == false), return the largest contiguous number of pages that
//    exist
static Buffer alloc_raw(s64 count, bool exact);

// get physical address from kernel address
u64 physical_address(const void *ptr) {
  u64 address = (u64)ptr;
  assert(address >= MEMORY__KERNEL_SPACE_BEGIN);

  return address - MEMORY__KERNEL_SPACE_BEGIN;
}

// get kernel address from physical address
u64 kernel_address(u64 address) {
  assert(address < MEMORY__KERNEL_SPACE_BEGIN);

  return address + MEMORY__KERNEL_SPACE_BEGIN;
}

void *kernel_ptr(u64 address) {
  assert(address < MEMORY__KERNEL_SPACE_BEGIN);

  return (void *)(address + MEMORY__KERNEL_SPACE_BEGIN);
}

static inline BuddyInfo buddy_info(s64 page, s64 class) {
  assert(is_aligned(page, 1 << class));
  s64 buddy = page ^ (S64(1) << class);
  s64 index = page >> (class + 1);
  return (BuddyInfo){.buddy = buddy, .bitset_index = index};
}

static inline MemSizeFormat mem_fmt(s64 size) {
  static const char *const sizename[] = {"", " Kb", " Mb", " Gb"};
  u8 class = 0;
  REPEAT(3) {
    if (size < 1024)
      break;
    class += 1;
    size /= 1024;
  }

  return (MemSizeFormat){.size = size, .suffix = sizename[class]};
}

void memory__init() {
  // Calculation described in bootboot specification
  MMap mmap = {.data = &bb.mmap, .count = (bb.size - 128) / 16};
  const MMapEnt *last_ent = &mmap.data[mmap.count - 1];
  const u64 end_address = MMapEnt_Ptr(last_ent) + MMapEnt_Size(last_ent);
  const u64 end_page = align_up(end_address, _4KB) / _4KB;

  FOR(mmap) {
    static const char *const typename[] = {"Used", "Free", "ACPI", "MMIO"};
    const char *const mtype = typename[MMapEnt_Type(it)];
    MemSizeFormat size_fmt = mem_fmt((s64)MMapEnt_Size(it));
    log_fmt("ENT: type=%f size=%f%f", mtype, size_fmt.size, size_fmt.suffix);
  }
  log();

  // sort the entries so that the free ones are first
  SLOW_SORT(mmap) {
    const bool swap =
        (MMapEnt_Type(left) != MMAP_FREE) & (MMapEnt_Type(right) == MMAP_FREE);
    if (swap)
      SWAP(left, right);
  }

  // remove weird bit stuff that BOOTBOOT does for the free entries
  FOR(mmap) {
    if (!MMapEnt_IsFree(it)) {
      mmap.count = index;
      break;
    }

    it->size = MMapEnt_Size(it);
  }

  // Hacky solution to quickly get everything into a higher-half kernel
  UNSAFE_HACKY_higher_half_init();

  log_fmt("higher-half addressing INIT COMPLETE");

  // Build basic buddy system structure
  const u64 buddy_max = align_up(end_address, _4KB << CLASS_COUNT);
  const s64 buddy_end_page = buddy_max / _4KB;

  u64 *usable_pages_data = alloc_from_entries(mmap, (end_page - 1) / 8 + 1, 8);
  MemGlobals.usable_pages = BitSet__from_raw(usable_pages_data, buddy_end_page);
  BitSet__set_all(MemGlobals.usable_pages, false);

  u64 *free_pages_data = alloc_from_entries(mmap, (end_page - 1) / 8 + 1, 8);
  MemGlobals.free_pages = BitSet__from_raw(free_pages_data, buddy_end_page);
  BitSet__set_all(MemGlobals.free_pages, false);

  FOR_PTR(MemGlobals.classes, CLASS_COUNT - 1, info, class) {
    const s64 bitset_size =
        align_up(buddy_info(buddy_end_page, class).bitset_index, 8);
    u64 *const data = alloc_from_entries(mmap, bitset_size / 8, 8);
    info->buddies = BitSet__from_raw(data, bitset_size);
    BitSet__set_all(info->buddies, false);
  }

  s64 available_memory = 0;
  FOR(mmap, entry) {
    u64 begin = align_up(entry->ptr, _4KB);
    u64 end = align_down(entry->ptr + entry->size, _4KB);
    s64 begin_page = begin / _4KB;
    s64 end_page = end / _4KB;
    s64 size = S64(max(end, begin) - begin);

    available_memory += size;

    BitSet__set_range(MemGlobals.usable_pages, begin_page, end_page, true);
    release_pages(kernel_ptr(begin), size / _4KB);
  }

  assert(available_memory == MemGlobals.free_memory);
  validate_heap();

  log_fmt("global allocator INIT_COMPLETE");

  // Build a new page table using functions that assume higher-half kernel
  PageTable4 *old = get_page_table();
  PageTable4 *new = zeroed_pages(1);
  assert(new);

  // Map higher half code
  const void *const target = kernel_ptr(0);
  bool res = map_region(new, (u64)target, target, (s64)end_page, PTE_KERNEL);
  assert(res);

  const u8 *code_ptr = &code_begin, *code_end_ptr = &code_end,
           *bss_end_ptr = &bss_end;
  const s64 code_size = S64(code_end_ptr - code_ptr),
            bss_size = S64(bss_end_ptr - code_end_ptr);

  // Map kernel code to address listed in the linker script
  res = copy_mapping(new, old, (u64)code_ptr, code_size / _4KB, PTE_KERNEL_EXE);
  assert(res);

  // Map BSS data
  res = copy_mapping(new, old, (u64)code_end_ptr, bss_size / _4KB, PTE_KERNEL);
  assert(res);

  // Map Bootboot struct, as described in linker script
  res = copy_mapping(new, old, (u64)&bb, 1, PTE_KERNEL);
  assert(res);

  // Map Environment data
  res = copy_mapping(new, old, (u64)&environment, 1, PTE_KERNEL);
  assert(res);

  // Map bootboot kernel stack
  res = copy_mapping(new, old, 0xFFFFFFFFFFFFF000, 1, PTE_KERNEL);
  assert(res);

  res = copy_mapping(new, old, (u64)&fb, align_up(bb.fb_size, _4KB) / _4KB,
                     PTE_KERNEL);
  assert(res);

  // Make sure BSS data stays up-to-date (because it includes MemGlobals)
  set_page_table(new);
  validate_heap();

  destroy_bootboot_table(old);
  validate_heap();

  // NOTE: this leaks intentionally. The GDT and IDT need to exist until
  // shutdown, at which time it does not matter whether they are freed.
  InitAlloc = Bump__new(2);

  log_fmt("memory INIT_COMPLETE");
}

static void *alloc_from_entries(MMap mmap, s64 _size, s64 _align) {
  assert(_size > 0 && _align >= 0);

  u64 align = max((u64)_align, 1);
  u64 size = align_up((u64)_size, align);

  FOR(mmap) {
    u64 aligned_ptr = align_up(it->ptr, align);
    u64 aligned_size = it->size + it->ptr - aligned_ptr;
    if (aligned_size < size)
      continue;

    it->ptr = aligned_ptr + size;
    it->size = aligned_size - size;
    return (void *)(aligned_ptr + MEMORY__KERNEL_SPACE_BEGIN);
  }

  assert(false);
}

static inline FreeBlock *find_block(FreeBlock *target) {
  FOR_PTR(MemGlobals.classes, CLASS_COUNT, info, class) {
    FreeBlock *block = info->freelist;
    while (block != NULL) {
      if (block == target)
        return block;
      block = block->next;
    }
  }

  return NULL;
}

static void *pop_freelist(s64 class) {
  assert(class < CLASS_COUNT);

  ClassInfo *info = &MemGlobals.classes[class];
  FreeBlock *block = info->freelist;

  assert(block != NULL);
  assert(block->class == class, "block had class %f in freelist of class %f",
         block->class, class);
  assert(block->prev == NULL);

  info->freelist = block->next;
  if (info->freelist != NULL)
    info->freelist->prev = NULL;

  return block;
}

static inline void remove_from_freelist(s64 page, s64 class) {
  assert(class < CLASS_COUNT);
  assert(is_aligned(page, 1 << class));

  FreeBlock *block = kernel_ptr(U64(page) * _4KB);
  assert(block->class == class);

  ClassInfo *info = &MemGlobals.classes[class];
  FreeBlock *prev = block->prev, *next = block->next;
  if (next != NULL)
    next->prev = prev;
  if (prev != NULL) {
    prev->next = next;
  } else {
    if (info->freelist != block) {
      log_fmt("in size class %f: freelist=%f and block=%f", class,
              (u64)info->freelist, (u64)block);
      s64 counter = 0;
      for (FreeBlock *i = info->freelist; i != NULL; i = i->next, counter++) {
        log_fmt("%f: block=%f", counter, (u64)i);
      }

      assert(false);
    }

    info->freelist = next;
  }
}

static inline void add_to_freelist(s64 page, s64 class) {
  assert(class < CLASS_COUNT);
  assert(is_aligned(page, 1 << class));

  FreeBlock *block = kernel_ptr(U64(page) * _4KB);
  ClassInfo *info = &MemGlobals.classes[class];

  block->class = class;
  block->prev = NULL;
  block->next = info->freelist;

  if (info->freelist != NULL)
    info->freelist->prev = block;
  info->freelist = block;
}

void validate_heap(void) {
  bool success = true;
  s64 calculated_free_memory = 0;
  for (s64 i = 0; i < CLASS_COUNT; i++) {
    s64 size = (s64)((1 << i) * _4KB);
    FreeBlock *block = MemGlobals.classes[i].freelist;
    for (; block != NULL; block = block->next) {
      if (block->class != i) {
        log_fmt("block class = %f but was in class %f", block->class, i);
        success = false;
      }

      calculated_free_memory += size;
    }
  }

  if (calculated_free_memory != MemGlobals.free_memory) {
    log_fmt("calculated was %f but free_memory was %f", calculated_free_memory,
            MemGlobals.free_memory);
    success = false;
  }

  assert(success);
}

void *zeroed_pages(s64 count) {
  void *data = alloc_raw(count, true).data;
  ensure(data) return NULL;

  memset(data, 0, count * _4KB);
  return data;
}

Buffer try_raw_pages(s64 count) { return alloc_raw(count, false); }

void *raw_pages(s64 count) { return alloc_raw(count, true).data; }

static Buffer alloc_raw(s64 count, bool exact) {
  Buffer buf = (Buffer){.data = NULL, .count = 0};
  if (count <= 0)
    return buf;

  const s64 min_class = smallest_greater_power2(count);
  s64 class = -1;
  NAMED_BREAK(found_class) {
    RANGE(min_class, CLASS_COUNT, current) {
      if (MemGlobals.classes[current].freelist) {
        class = current;
        break(found_class);
      }
    }

    // Could'nt allocate exactly the required amount
    if (exact)
      return buf;

    for (s64 it = min_class - 1; it >= 0; it--) {
      if (MemGlobals.classes[it].freelist) {
        count = 1 << it;
        class = it;
        break(found_class);
      }
    }

    // There's nothing left lol
    return buf;
  }

  const s64 size = count * _4KB;
  MemGlobals.free_memory -= size;

  buf.data = pop_freelist(class);
  buf.count = size;

  const u64 addr = physical_address(buf.data);
  const s64 begin = addr / _4KB, end = begin + count;
  assert(BitSet__get_all(MemGlobals.usable_pages, begin, end));
  assert(BitSet__get_all(MemGlobals.free_pages, begin, end));
  BitSet__set_range(MemGlobals.free_pages, begin, end, false);

  if (class != CLASS_COUNT - 1) {
    s64 index = buddy_info(begin, class).bitset_index;
    assert(BitSet__get(MemGlobals.classes[class].buddies, index));
    BitSet__set(MemGlobals.classes[class].buddies, index, false);
  }

  DECLARE_SCOPED(s64 remaining = count, page = begin)
  for (s64 i = class; remaining > 0 && i > 0; i--) {
    const s64 child_class = i - 1;
    ClassInfo *const info = &MemGlobals.classes[child_class];
    const s64 child_size = 1 << child_class;
    const s64 index = buddy_info(page, child_class).bitset_index;
    assert(!BitSet__get(info->buddies, index));

    if (remaining > child_size) {
      remaining -= child_size;
      page += child_size;
      continue;
    }

    add_to_freelist(page + child_size, child_class);
    BitSet__set(info->buddies, index, true);

    if (remaining == child_size)
      break;
  }

  return buf;
}

void release_pages(void *data, s64 count) {
  assert(data != NULL);
  const u64 addr = physical_address(data);
  assert(addr == align_down(addr, _4KB));

  const s64 begin = addr / _4KB, end = begin + count;
  assert(BitSet__get_all(MemGlobals.usable_pages, begin, end));
  assert(!BitSet__get_any(MemGlobals.free_pages, begin, end));

  RANGE(begin, end, page) {
    // TODO should probably do some math here to not have to iterate over every
    // page in data
    FOR_PTR(MemGlobals.classes, CLASS_COUNT - 1, info, class) {
      assert(is_aligned(page, 1 << class));
      const BuddyInfo buds = buddy_info(page, class);

      const bool buddy_is_free = BitSet__get(info->buddies, buds.bitset_index);
      BitSet__set(info->buddies, buds.bitset_index, !buddy_is_free);

      if (!buddy_is_free) {
        add_to_freelist(page, class);
        continue(page);
      }

      remove_from_freelist(buds.buddy, class);
      page = min(page, buds.buddy);
    }

    assert(is_aligned(page, 1 << (CLASS_COUNT - 1)));
    add_to_freelist(page, CLASS_COUNT - 1);
  }

  MemGlobals.free_memory += count * _4KB;
  BitSet__set_range(MemGlobals.free_pages, begin, end, true);
}

void unsafe_mark_memory_usability(const void *data, s64 count, bool usable) {
  assert(data != NULL);
  const u64 addr = physical_address(data);
  assert(addr == align_down(addr, _4KB));

  const s64 begin_page = addr / _4KB, end_page = begin_page + count;

  const bool any_are_free =
      BitSet__get_any(MemGlobals.free_pages, begin_page, end_page);
  assert(!any_are_free,
         "if you're marking memory usability, the marked pages can't be free");

  const s64 existing_pages =
      BitSet__get_count(MemGlobals.usable_pages, begin_page, end_page);
  BitSet__set_range(MemGlobals.usable_pages, begin_page, end_page, usable);
}
